### PR TITLE
[backup] db-restore: dry run mode

### PR DESCRIPTION
**Pull request by msmouse**
## Motivation

dry run restore on all backups in a storage is equivalent to verifying the entire backup data.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

existing coverage for the non dry-run mode.
gonna add smoke test for the verify mode.

### CODE REVIEW

File: storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs [hunk 1]
```diff
@@ -138,3 +173,122 @@ impl EpochEndingRestoreController {
...
+            epoch,
+        );
+        if epoch == 0 {
+            ensure!(
+                li_with_sigs.ledger_info() == &self.epoch_endings[0],
+                "Genesis epoch LedgerInfo info doesn't match.",
+            );
+            Ok(())
+        } else {
+            self.epoch_endings[epoch as usize - 1]
```
Comment (lightmark) [inline storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs:211]: I feel it is better to have genesis_ledger_info at index 0.

File: storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs [hunk 2]
```diff
@@ -138,3 +173,122 @@ impl EpochEndingRestoreController {
...
+                );
+                ensure!(
+                    li.ends_epoch(),
+                    "LedgerInfo is not one at an epoch ending. epoch: {}",
+                    li.epoch(),
+                );
+                next_epoch += 1;
+            }
+
+            previous_li = Some(lis.last().unwrap().clone());
```
Comment (lightmark) [inline storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs:296]:
```suggestion
            previous_li = Some(lis.last().expect("non-emptiness verified").clone());
```

File: storage/backup/backup-cli/src/bin/db-restore.rs [hunk 3]
```diff
@@ -86,6 +86,7 @@ async fn main_impl() -> Result<()> {
                 global_opt,
                 storage.init_storage().await?,
                 restore_handler,
+                None,
```
Comment (lightmark) [inline storage/backup/backup-cli/src/bin/db-restore.rs]: comment for `None`?

File: storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs [hunk 4]
```diff
@@ -147,9 +150,17 @@ impl EpochEndingRestoreController {
 
             // write to db
             if !lis.is_empty() {
-                self.restore_handler.save_ledger_infos(&lis)?;
-                EPOCH_ENDING_EPOCH.set(lis.last().unwrap().ledger_info().epoch() as i64);
-                EPOCH_ENDING_VERSION.set(lis.last().unwrap().ledger_info().version() as i64);
+                match self.run_mode.as_ref() {
+                    RestoreRunMode::Restore { restore_handler } => {
+                        restore_handler.save_ledger_infos(&lis)?;
+                        EPOCH_ENDING_EPOCH.set(lis.last().unwrap().ledger_info().epoch() as i64);
```
Comment (lightmark) [inline storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs:156]: avoid `unwrap()`, at least using `expect`?

**Review by lightmark — APPROVED** [review]
The first commit has no commit message so I am not sure what's its goal.

File: storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs (hunk 4 above)
Comment (msmouse) [reply storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs:156]: K. updated in the second diff in the stack. Although I think the coding guideline around absolutely no unwrap() doesn't really make sense. 🏃‍♀️

File: storage/backup/backup-cli/src/bin/db-restore.rs (hunk 3 above)
Comment (msmouse) [reply storage/backup/backup-cli/src/bin/db-restore.rs]: done

**Comment by msmouse** [comment]
> The first commit has no commit message so I am not sure what's its goal.

updated, it's to make the test in the following commit pass (epoch ending ledger infos now have real signatures that can verify against the previous epoch state)

**Pull request closed without merging**